test/integration/volume/attach_detach_test.go

```diff
@@ -138,6 +143,110 @@ var defaultTimerConfig = attachdetach.TimerConfig{
 	DesiredStateOfWorldPopulatorListPodsRetryDuration: 3 * time.Second,
 }
 
+// TestPodTerminationWithNodeOOSDetach integration test verifies that if `out-of-service` taint is applied to the node,
+// which is shutdown non gracefully, then all the pods will immediately get terminated and volume be immediately detached
+// without waiting for the default time out period.
+func TestPodTerminationWithNodeOOSDetach(t *testing.T) {
+	// Disable ServiceAccount admission plugin as we don't have serviceaccount controller running.
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins=ServiceAccount"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+	namespaceName := "test-volume-terminating-detach"
+	nodeName := "node-sandbox"
+	testClient, ctrl, pvCtrl, gcCtrl, informers := createAdClients(t, server, defaultSyncPeriod, defaultTimerConfig)
+	ns := framework.CreateNamespaceOrDie(testClient, namespaceName, t)
+	defer framework.DeleteNamespaceOrDie(testClient, ns, t)
+
+	// create a node
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+			Annotations: map[string]string{
+				util.ControllerManagedAttachAnnotation: "true",
+			},
+		},
+	}
+	node, err := testClient.CoreV1().Nodes().Create(context.TODO(), node, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to created node : %v", err)
+	}
+
+	// create fake pods with volumes
+	pod := fakePodWithVol(namespaceName)
+	if _, err := testClient.CoreV1().Pods(ns.Name).Create(context.TODO(), pod, metav1.CreateOptions{}); err != nil {
+		t.Errorf("Failed to create pod : %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go informers.Core().V1().Nodes().Informer().Run(ctx.Done())
+	podInformer := informers.Core().V1().Pods().Informer()
+	// start the informer
+	informers.Start(ctx.Done())
+	informers.WaitForCacheSync(ctx.Done())
+
+	// run the controllers
+	go ctrl.Run(ctx)
+	go pvCtrl.Run(ctx)
+	go gcCtrl.Run(ctx)
+
+	waitToObservePods(t, podInformer, 1)
+
+	// wait for volume to be attached
+	for i := 0; i < 10; i++ {
+		node, err = testClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("Failed to get the node : %v", err)
+		}
+		if len(node.Status.VolumesAttached) > 1 {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+	if len(node.Status.VolumesAttached) < 1 {
+		t.Logf("failed to attach volume for pod %s on node %s", pod.Name, node.Name)
```

Comment:
we have already libraries for doing these async checks

https://github.com/kubernetes/kubernetes/blob/854d0e7fc8e167785fc38dd90f88275f3228d3e3/test/integration/apiserver/watchcache_test.go#L78-L83

Reply:
oh, you already using them below, why do you implement the loop to check here differently?

Reply:
I will fix it. Missed this one. Thanks